File: frontend/src/types/course.ts
export interface Category {
  id: number | string;
  name: string;
  slug?: string;
  description?: string;
}

export interface Lecture {
  id: number | string;
  title: string;
  description?: string;
  durationMinutes?: number;
  durationSeconds?: number | null;
  orderIndex: number;
  type?: "VIDEO" | "PDF" | "TEXT" | "QUIZ" | string;
  videoUrl?: string | null;
  pdfUrl?: string | null;
  content?: string | null;
  contentHtml?: string;
  contentUrl?: string;
  thumbnailUrl?: string | null;
  assessmentId?: number | string;
  quizQuestionCount?: number;
  preview?: boolean;
  isPreview?: boolean;
  isLocked?: boolean;
}

export interface CourseSection {
  id: number | string;
  title: string;
  description?: string;
  orderIndex: number;
  lectures: Lecture[];
}

export interface Course {
  id: number | string;
  title: string;
  slug: string;
  description: string;
  thumbnailUrl?: string;
  durationMinutes: number;
  difficulty: "BEGINNER" | "INTERMEDIATE" | "ADVANCED" | string;
  published: boolean;
  category: Category;
  categoryId?: number | string;
  categoryName?: string;
  categoryDescription?: string;
  targetAudience?: string[];
  sections?: CourseSection[];
  enrolled?: boolean;
}

File: itas_backend/src/main/java/com/aauelknight/itas_backend/courses/entity/Course.java
package com.aauelknight.itas_backend.courses.entity;

import jakarta.persistence.CascadeType;
import jakarta.persistence.CollectionTable;
import jakarta.persistence.Column;
import jakarta.persistence.ElementCollection;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OrderBy;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import jakarta.persistence.Transient;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.Set;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.BatchSize;
import com.aauelknight.itas_backend.auth.entity.User;
@Entity
@Table(name = "courses")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Course {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 200)
    private String title;

    @Column(nullable = false, unique = true, length = 200)
    private String slug;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private CourseDifficulty difficulty;

    @Column(name = "duration_minutes")
    private Integer durationMinutes;

    @Column(name = "thumbnail_url", nullable = true, columnDefinition = "TEXT")
    private String thumbnailUrl;

    @Enumerated(EnumType.STRING)
    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "course_audience", joinColumns = @JoinColumn(name = "course_id"))
    @Column(name = "audience")
    @Builder.Default
    private Set<AudienceType> targetAudience = new HashSet<>(Set.of(AudienceType.ALL));

    @Column(nullable = false)
    private boolean published;

    @Column(nullable = false, length = 50)
    @Builder.Default
    private String status = "DRAFT";

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @Column(name = "archived_at")
    private LocalDateTime archivedAt;

    @Transient
    private String searchVector;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "archived_by")
    private User archivedBy;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    @OneToMany(mappedBy = "course", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("orderIndex ASC")
    @BatchSize(size = 50)
    @Builder.Default
    private Set<CourseSection> sections = new LinkedHashSet<>();

    @PrePersist
    public void onCreate() {
        LocalDateTime now = LocalDateTime.now();
        createdAt = now;
        updatedAt = now;
        if (status == null || status.isBlank()) {
            status = published ? "PUBLISHED" : "DRAFT";
        }
    }

    @PreUpdate
    public void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

